components/modules/mqtt.c

```diff
@@ -164,572 +156,633 @@ static esp_err_t mqtt_event_handler(esp_mqtt_event_handle_t event)
+
+    NODE_DBG("MQTT client garbage collected\n");
+    return 0;
 }
 
 // Lua: mqtt.Client(clientid, keepalive[, username, password, cleansession])
-static int mqtt_new( lua_State* L )
-{
-	const char * clientid = NULL;
-	clientid = luaL_checkstring( L, 1 );
-	NODE_DBG("MQTT client id %s\n", clientid);
-
-	esp_mqtt_lua_client_config_t * mqtt_cfg = (esp_mqtt_lua_client_config_t *) malloc(sizeof(esp_mqtt_lua_client_config_t));
-	memset(mqtt_cfg, 0, sizeof(esp_mqtt_lua_client_config_t));
-	mqtt_cfg->config.host = mqtt_cfg->host;
-	mqtt_cfg->config.uri = mqtt_cfg->uri;
-	mqtt_cfg->config.client_id = mqtt_cfg->client_id;
-	mqtt_cfg->config.username = mqtt_cfg->username;
-	mqtt_cfg->config.password = mqtt_cfg->password;
-	mqtt_cfg->config.lwt_topic = mqtt_cfg->lwt_topic;
-	mqtt_cfg->config.lwt_msg = mqtt_cfg->lwt_msg;
-
-	mqtt_cfg->config.event_handle = mqtt_event_handler;
-
-	strncpy(mqtt_cfg->client_id, clientid, MQTT_MAX_CLIENT_LEN);
-	mqtt_cfg->config.keepalive = luaL_checkinteger( L, 2 );
-
-	int n = 2;
-	if( lua_isstring(L, 3) )
-	{
-		strncpy( mqtt_cfg->username, luaL_checkstring( L, 3 ), MQTT_MAX_USERNAME_LEN);
-		n++;
-	}
-
-	if( lua_isstring(L, 4) )
-	{
-		strncpy(mqtt_cfg->password, luaL_checkstring( L, 4 ), MQTT_MAX_PASSWORD_LEN);
-		n++;
-	}
-
-	if( lua_isnumber(L, 5) )
-	{
-		mqtt_cfg->config.disable_clean_session = (luaL_checknumber( L, 5 ) == 0);
-		n++;
-	}
-	lua_pop( L, n ); //remove parameters
-
-	lua_newtable( L );
-	NODE_DBG("New MQTT table at stack pos %d\n", lua_gettop(L));
-
-	lua_pushlightuserdata( L, mqtt_cfg );
-	lua_setfield( L, -2, "_settings" );  // set t["_mqtt"] = client
-
-	luaL_getmetatable( L, "mqtt.mt" );
-	lua_setmetatable( L, -2 );
-
-	hConn  = task_get_id(_connected_cb);
-	hOff   = task_get_id(_disconnected_cb);
-	hPub   = task_get_id(_publish_cb);
-	hSub   = task_get_id(_subscribe_cb);
-	hUnsub = task_get_id(_unsubscribe_cb);
-	hData  = task_get_id(_data_cb);
-
-	NODE_DBG("conn %d, off %d, pub %d, sub %d, data %d\n", hConn, hOff, hPub, hSub, hData);
-	return 1; //leave table on top of the stack
+// mqtt_new creates a new instance of our mqtt userdata lua object
+static int mqtt_new(lua_State* L) {
+    //create a new lua userdata object and initialize to 0.
+    mqtt_context_t* mqtt_context = (mqtt_context_t*)lua_newuserdata(L, sizeof(mqtt_context_t));
+    memset(mqtt_context, 0, sizeof(mqtt_context_t));
+
+    // initialize all callbacks to LUA_NOREF, indicating they're unset.
+    for (int i = 0; i < sizeof(mqtt_context->event_cb) / sizeof(lua_ref_t); i++) {
+        mqtt_context->event_cb[i] = LUA_NOREF;
+    }
+
+    // keep a weak reference to our userdata object so we can pass it as a parameter to user callbacks
+    lua_pushvalue(L, -1);
+    mqtt_context->self = luaX_weak_ref(L);
+
+    // allocate a pointer that will be used to link the mqtt client user context to this lua userdata object
+    mqtt_context->pcontext = luaM_malloc(L, sizeof(mqtt_context_t*));
```

Comment:
This is the solution to a good one. Took me awhile to debug.

In certain occasions, for example if the userdata object is garbage collected while mqtt_client has still a pending operation it could happen that a late event is fired referencing an already collected object, causing a reboot or heap corruption. As a matter of fact, calling the `destroy` function to free the esp_mqtt_client may actually fire disconnect events that are too late to the party.

This double indirection pointer allows the `__gc` function `mqtt_delete()` to set it to `NULL` so that if a rogue event is fired, the `mqtt_event_handler()` function can ignore the event.

The esp_mqtt_client library takes a `user_context` pointer on initialization that cannot be changed later. By having `user_context` point to a pointer instead, I can change that second pointer to `NULL` to signal the event handler to ignore further events.

See `mqtt_delete()` and `mqtt_event_handler()` comments. Please let me know if it is clear enough.

Without this sort of approach, the following fails:

```lua
client = mqtt.Client("test", 100)
client:connect(
    "1.2.3.4",
    1883,
    0,
    function()
        print("connected")
    end,
    function()
        print("failed")
    end
)
client = nil -- this can cause client to be garbage collected before the above events are fired.
```

Note that if the lua user closed the connection politely or was diligent with his references, this would not happen. Dereferencing the client is a brutal way to disconnect, but we shouldn't punish the user with random reboots when writing correct lua.

Reply:
Never mind. I found a solution to the above problem without an allocated pointer. I also refactored the code and simplified it even further.